### PR TITLE
fix: use new URL instead of url.parse

### PR DESCRIPTION
**Pull request by dead-horse**
##### Checklist

- [ ] `npm test` passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows commit guidelines

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,7 +3,6 @@ language: node_js
 node_js:
   - '8'
   - '10'
-  - '12'
```
Comment (fengmk2) [inline .travis.yml]: keep node 12

File: app.js [hunk 2]
```diff
@@ -12,8 +13,16 @@ module.exports = app => {
     const origin = ctx.get('origin');
     if (!origin) return '';
 
-    const parsedUrl = url.parse(origin);
-    if (!ctx.isSafeDomain || ctx.isSafeDomain(parsedUrl.hostname) || ctx.isSafeDomain(origin)) {
+    if (!ctx.isSafeDomain) return origin;
```
Comment (fengmk2) [inline app.js]: typeof ctx.isSafeDomain !== 'function' will be better.

File: app.js [hunk 3]
```diff
@@ -12,8 +13,16 @@ module.exports = app => {
...
+
+    let parsedUrl;
+    try {
+      parsedUrl = new URL(origin);
+    } catch (err) {
+      return '';
+    }
+
+    if (ctx.isSafeDomain(parsedUrl.hostname) || ctx.isSafeDomain(origin)) {
```
Comment (fengmk2) [inline app.js:25]: 当时设计为什么会调用2次。。。有点奇怪。

Comment (dead-horse) [reply app.js:25]: https://github.com/eggjs/egg-cors/pull/21

Comment (dead-horse) [reply app.js:25]: 感觉这个兼容有点问题，是不是可以传 host 就好

**Review by fengmk2 — APPROVED** [review]

File: app.js (hunk 3 above)
Comment (dead-horse) [reply app.js:25]:
```
domainWhiteList: [
    '.eggjs.org',
    'https://a.com',
    'https://b.com:1234',
    'https://*.c.com',
  ],
```

额，是因为 domainWhiteList 的配置支持了 protocol + port 的形式

Comment (fengmk2) [reply app.js:25]: 应该是第二个参数传 origin

**Comment by dead-horse** [comment]
2.2.3

**Pull request merged**